### PR TITLE
Implementation of PD map to AF map conversion feature 

### PR DESCRIPTION
**Pull request by acanalpay**
Two buttons were added in the edit section of the menu and the right-click menu.  In the app-menu.js, code creates a string file from the visible map and sends it to the pd2af-webservice. pd2af-webservice prepares the format and sends it to the conversion server. After the conversion server's response, pd2af-webservice returns the appropriate response. Then in app-menu.js, the file is opened from the URL on the new Newt tab.

### CODE REVIEW

File: app/js/app-menu.js [hunk 1]
```diff
@@ -964,7 +971,133 @@ module.exports = function() {
...
+
+      var url = "/";
+      var retdata;
+      // pd2af returns filename and file url
+      $.ajax({
+        // After deploying Bridge Server (pd2af-webservice) write the bridge server's URL but leave the /convert
+        url: "https://pd2afwebservice.herokuapp.com/convert",
+        // url: "http://localhost:5555/convert",
+        type: "POST",
+        timeout: 300000,
```
Comment (hasanbalci) [inline app/js/app-menu.js:1027]: Does this mean to wait conversion service for 300 seconds in the worst case? Maybe we can do it 30 seconds.

File: app/js/app-menu.js [hunk 2]
```diff
@@ -964,7 +971,133 @@ module.exports = function() {
       var cy = chiseInstance.getCy();
 
       chiseInstance.highlightProcesses(cy.nodes(':selected'));
+      });
```
Comment (hasanbalci) [inline app/js/app-menu.js:974]: Fix indentation and add a new line after this.

File: app/js/app-menu.js [hunk 3]
```diff
@@ -964,7 +971,133 @@ module.exports = function() {
...
+        error: function (data) {
+          console.log(data);
+          chiseSpinner.endSpinner("layout-spinner");
+
+          promtErrorPD2AF.render(data);
+        }
+      });
+
     });
+
```
Comment (hasanbalci) [inline app/js/app-menu.js:1100]: Let's keep only one empty line here between blocks (currently three)

File: app/js/backbone-views.js [hunk 4]
```diff
@@ -3303,7 +3303,50 @@ var PromptInvalidFileView = Backbone.View.extend({
...
 });
+var PromptInvalidTypeWarning = Backbone.View.extend({
+  initialize: function () {
+    var self = this;
+    self.template = _.template($("#prompt-errorInvalidType-template").html());
+  },
+  render: function () {
+    var self = this;
+    self.template = _.template($("#prompt-errorInvalidType-template").html());
+    // console.log("RENDEEER");
```
Comment (hasanbalci) [inline app/js/backbone-views.js:3314]: Again, let's remove these commented statements.

File: package.json [hunk 5]
```diff
@@ -7,7 +7,8 @@
     "watch-js": "nodemon -e js --watch app --ignore app/bundle.js -x \"npm run build-bundle-js\"",
     "watch-server": "nodemon server.js",
     "release-build": "npm run build-bundle-js",
-    "debug-build": "parallelshell \"npm run watch-js\" \"npm run watch-server\""
+    "debug-build": "parallelshell \"npm run watch-js\" \"npm run watch-server\"",
+    "heroku-postbuild": "npm run build-bundle-js"
```
Comment (hasanbalci) [inline package.json:11]: I think this is Heroku related and can be removed.

File: package.json [hunk 6]
```diff
@@ -60,8 +62,11 @@
     "cypress": "^7.3.0",
     "cypress-file-upload": "^5.0.3",
     "cypress-real-events": "^1.3.0",
-    "express": "^4.14.0",
```
Comment (hasanbalci) [inline package.json:63]: Do we really need to move express from devDepenencies to dependencies?

File: package.json [hunk 7]
```diff
@@ -60,8 +62,11 @@
...
+    "body-parser": "*",
```
Comment (hasanbalci) [inline package.json:65]: I think it would be better to give a specific version.

File: package.json [hunk 8]
```diff
@@ -60,8 +62,11 @@
...
     "nodemon": "1.11.0",
     "parallelshell": "3.0.1"
+  },
+  "engines": {
```
Comment (hasanbalci) [inline package.json:69]: Let's remove this `engines` field.

File: app/js/app-menu.js [hunk 9]
```diff
@@ -964,7 +971,133 @@ module.exports = function() {
...
+    $("#highlight-errors-of-pd2af").click(function (e) {
+    
+      var chiseInstance = appUtilities.getActiveChiseInstance();
+      var chiseSpinner = appUtilities.getActiveChiseInstance();
+      var filename = document.getElementById('file-name').innerHTML;
+      var fExt = 'sbgn';
+      filename = filename.substring(0, filename.lastIndexOf('.')).concat(".").concat(fExt);
+
+      // Get the map from the canvas
+
```
Comment (hasanbalci) [inline app/js/app-menu.js:985]: Remove empty line.

File: app/js/app-menu.js [hunk 10]
```diff
@@ -602,12 +604,17 @@ module.exports = function() {
...
       // reset map name and description
       // default map name should be a string that contains the network id
-      currentGeneralProperties.mapName = appUtilities.getDefaultMapName(networkId);
-      currentGeneralProperties.mapDescription = appUtilities.defaultGeneralProperties.mapDescription;
-    
+      if(!currentGeneralProperties.mapPD2AFConverted) {// if the file don't have a specified map name from pd map to af map conversion
+        currentGeneralProperties.mapName = appUtilities.getDefaultMapName(networkId);
+        currentGeneralProperties.mapDescription = appUtilities.defaultGeneralProperties.mapDescription;
+      }
+      // We temporarly set inferNestingOnLoad to true to preserve the nest relation in the pd2af conversion
```
Comment (hasanbalci) [inline app/js/app-menu.js:613]: It seems that comment doesn't match with the action.

File: app/js/app-menu.js [hunk 11]
```diff
@@ -964,7 +971,133 @@ module.exports = function() {
...
+      }
+
+      chiseSpinner.startSpinner("layout-spinner");
+
+      var url = "/";
+      var retdata;
+      // pd2af returns filename and file url
+      $.ajax({
+        // After deploying Bridge Server (pd2af-webservice) write the bridge server's URL but leave the /convert
+        url: "https://pd2afwebservice.herokuapp.com/convert",
```
Comment (hasanbalci) [inline app/js/app-menu.js:1024]: Don't forget to change here after you make the service active in the public server.

**Review by hasanbalci — CHANGES_REQUESTED** [review]
I added some comments in terms of styling. I'll also test it when conversion service is available in public server.

File: app/js/app-menu.js [hunk 12]
```diff
@@ -964,7 +971,133 @@ module.exports = function() {
       var cy = chiseInstance.getCy();
 
       chiseInstance.highlightProcesses(cy.nodes(':selected'));
+      });
+    // PD map to AF map conversion 
+    $("#highlight-errors-of-pd2af").click(function (e) {
+    
+      var chiseInstance = appUtilities.getActiveChiseInstance();
+      var chiseSpinner = appUtilities.getActiveChiseInstance();
```
[the remark this answers is not in the document]
Comment (acanalpay) [reply app/js/app-menu.js]: Done

File: app/js/app-menu.js [hunk 13]
```diff
@@ -964,7 +971,133 @@ module.exports = function() {
...
+      var filename = document.getElementById('file-name').innerHTML;
+      var fExt = 'sbgn';
+      filename = filename.substring(0, filename.lastIndexOf('.')).concat(".").concat(fExt);
+
+      // Get the map from the canvas
+      
+      var cy = chiseInstance.getCy();
+      var nodes = cy.nodes().filter( function( node ) {
+        // return !chiseInstance.elementUtilities.isSIFNode( node );
```
[the remark this answers is not in the document]
Comment (acanalpay) [reply app/js/app-menu.js]: Done

File: app/js/app-menu.js (hunk 1 above)
Comment (acanalpay) [reply app/js/app-menu.js]: It was a test for checking timeout. I decided to handle timeout in the bridge server. 
timeout deleted from .ajax

File: app/js/app-menu.js (hunk 2 above)
Comment (acanalpay) [reply app/js/app-menu.js]: Done

File: app/js/app-menu.js (hunk 3 above)
Comment (acanalpay) [reply app/js/app-menu.js:1099]: Done

File: app/js/backbone-views.js (hunk 4 above)
Comment (acanalpay) [reply app/js/backbone-views.js]: Done

File: package.json (hunk 6 above)
Comment (acanalpay) [reply package.json:63]: For Heroku, we have to. I'll move to devDependicies after resolving issues

File: package.json (hunk 8 above)
Comment (acanalpay) [reply package.json]: Done

File: app/js/app-menu.js (hunk 10 above)
Comment (acanalpay) [reply app/js/app-menu.js:613]: Done

Comment (hasanbalci) [reply app/js/app-menu.js:613]: I didn't understand what is done. Comment says "We temporarly set inferNestingOnLoad to true" but below code sets it to false.

File: app/js/app-menu.js (hunk 11 above)
Comment (acanalpay) [reply app/js/app-menu.js:1021]: This section is unnecessary. I'll delete it on the next commit

File: app/js/app-menu.js (hunk 10 above)
Comment (acanalpay) [reply app/js/app-menu.js:613]: This section is unnecessary. I'll delete it on the next commit

File: app/js/app-menu.js (hunk 11 above)
Comment (acanalpay) [reply app/js/app-menu.js]: Changed to http://localhost:4000/convert

File: package.json (hunk 5 above)
Comment (acanalpay) [reply package.json]: Deleted heroku-postbuild

File: package.json (hunk 6 above)
Comment (acanalpay) [reply package.json:63]: Moved express back to devDependicies

**Review by hasanbalci — APPROVED** [review]

**Pull request merged**